File: src/main/java/forpleuvoir/hiirosakura/client/feature/tooltip/Tooltip.java
package forpleuvoir.hiirosakura.client.feature.tooltip;

import com.google.common.collect.Lists;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.reflect.TypeToken;
import forpleuvoir.hiirosakura.client.config.Configs;
import forpleuvoir.hiirosakura.client.config.base.AbstractHiiroSakuraData;
import forpleuvoir.hiirosakura.client.util.HSLogger;
import forpleuvoir.hiirosakura.client.util.JsonUtil;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.text.LiteralText;
import net.minecraft.text.Text;

import java.util.*;

import static forpleuvoir.hiirosakura.client.util.PlayerHeadUtil.getSkullOwner;

/**
 * 工具提示
 *
 * @author forpleuvoir
 * <p>#project_name hiirosakura
 * <p>#package forpleuvoir.hiirosakura.client.feature.tooltip
 * <p>#class_name Tooltip
 * <p>#create_time 2021/6/17 0:55
 */
public class Tooltip extends AbstractHiiroSakuraData {

    private transient static final HSLogger log = HSLogger.getLogger(Tooltip.class);
    public Map<String, List<String>> datas = new HashMap<>();

    public Tooltip() {
        super("tooltip");
    }

    private void add(String key, String... tooltip) {
        if (tooltip == null || tooltip.length == 0) return;
        List<String> tooltips = Lists.newArrayList();
        Arrays.stream(tooltip).forEach(s -> tooltips.add(s.replace("&", "§")));
        if (datas.containsKey(key)) {
            datas.get(key).addAll(tooltips);
        } else {
            datas.put(key, tooltips);
        }
        this.onValueChanged();
    }

    public void add(ItemStack item, String... tooltip) {
        add(getKey(item), tooltip);
    }

    public String remove(ItemStack item, int index) {
        return remove(getKey(item), index);
    }

    private String remove(String key, int index) {
        if (datas.containsKey(key)) {
            StringBuilder s = new StringBuilder();
            if (index < 0) {
                for (String s1 : datas.get(key)) {
                    s.append(s1);
                    s.append(",");
                }
                s.deleteCharAt(s.length()-1);
                datas.remove(key);
            } else {
                try {
                    s = new StringBuilder(datas.get(key).get(index));
                    datas.get(key).remove(index);
                    if(datas.get(key).size()==0){
                        datas.remove(key);
                    }
                } catch (IndexOutOfBoundsException e) {
                    return null;
                }
            }
            this.onValueChanged();
            return s.toString();
        }
        return null;
    }

    public List<Text> getTooltip(ItemStack stack) {
        var list = new ArrayList<Text>();
        if (Configs.Toggles.SHOW_TOOLTIP.getBooleanValue()) {
            String key = getKey(stack);
            if (datas.containsKey(key)) {
                var tips = datas.get(key);
                tips.forEach(tip -> list.add(new LiteralText(tip)));
            }
        }
        return list;
    }

    @Override
    public void setValueFromJsonElement(JsonElement element) {
        try {
            if (element.isJsonObject()) {
                JsonObject object = element.getAsJsonObject();
                Map<String, List<String>> data = JsonUtil.gson
                        .fromJson(object, new TypeToken<Map<String, List<String>>>() {
                        }.getType());
                datas.clear();
                datas.putAll(data);
            } else {
                log.warn("{}无法从JsonElement{}中读取数据", this.getName(), element);
            }
        } catch (Exception e) {
            log.warn("{}无法从JsonElement{}中读取数据", this.getName(), element, e);
        }
    }

    @Override
    public JsonElement getAsJsonElement() {
        return JsonUtil.gson.toJsonTree(datas);
    }

    public static String getKey(ItemStack stack) {
        String value = "";
        Item item = stack.getItem();
        String name = item.getName(stack).getString();
        String cName = stack.getName().getString();
        if (stack.getItem().equals(Items.PLAYER_HEAD)) {
            value = ":" + getSkullOwner(stack);
        } else if (!name.equals(cName)) {
            value = "#" + cName;
        }
        return item.getTranslationKey(stack) + value;
    }


}

File: src/main/kotlin/forpleuvoir/hiirosakura/client/feature/tooltip/Tooltip.kt
package forpleuvoir.hiirosakura.client.feature.tooltip

import com.google.common.collect.Lists
import com.google.gson.JsonElement
import com.google.gson.reflect.TypeToken
import forpleuvoir.hiirosakura.client.config.Configs
import forpleuvoir.hiirosakura.client.config.base.AbstractHiiroSakuraData
import forpleuvoir.hiirosakura.client.util.HSLogger.Companion.getLogger
import forpleuvoir.hiirosakura.client.util.JsonUtil
import forpleuvoir.hiirosakura.client.util.PlayerHeadUtil
import forpleuvoir.ibuki_gourd.mod.utils.IbukiGourdLang
import forpleuvoir.ibuki_gourd.utils.text
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.text.Text
import java.util.*

/**
 * 工具提示
 *
 * @author forpleuvoir
 *
 * 项目名 hiirosakura
 *
 * 包名 forpleuvoir.hiirosakura.client.feature.tooltip
 *
 * 文件名 Tooltip
 *
 * 创建时间 2021/6/17 0:55
 */
class Tooltip : AbstractHiiroSakuraData("tooltip") {
	private val data: MutableMap<String, MutableList<String>> = HashMap()
	private fun add(key: String, vararg tooltip: String) {
		if (tooltip.isEmpty()) return
		val tooltips: MutableList<String> = Lists.newArrayList()
		Arrays.stream(tooltip).forEach { s: String -> tooltips.add(s.replace("&", "§")) }
		if (data.containsKey(key)) {
			data[key]!!.addAll(tooltips)
		} else {
			data[key] = tooltips
		}
		onValueChanged()
	}

	fun add(item: ItemStack, vararg tooltip: String) {
		add(getKey(item), *tooltip)
	}

	fun remove(item: ItemStack, index: Int): String? {
		return remove(getKey(item), index)
	}

	private fun remove(key: String, index: Int): String? {
		if (data.containsKey(key)) {
			var s = StringBuilder()
			if (index < 0) {
				for (s1 in data[key]!!) {
					s.append(s1)
					s.append(",")
				}
				s.deleteCharAt(s.length - 1)
				data.remove(key)
			} else {
				try {
					s = StringBuilder(data[key]!![index])
					data[key]!!.removeAt(index)
					if (data[key]!!.size == 0) {
						data.remove(key)
					}
				} catch (e: IndexOutOfBoundsException) {
					return null
				}
			}
			onValueChanged()
			return s.toString()
		}
		return null
	}

	fun getTooltip(stack: ItemStack): List<Text> {
		val list = ArrayList<Text>()
		if (Configs.Toggles.SHOW_TOOLTIP.getValue()) {
			val key = getKey(stack)
			if (data.containsKey(key)) {
				val tips: List<String> = data[key]!!
                tips.forEach { list.add(it.text) }
			}
		}
		return list
	}

	override fun setValueFromJsonElement(jsonElement: JsonElement) {
		try {
			if (jsonElement.isJsonObject) {
				val obj = jsonElement.asJsonObject
				val data = JsonUtil.gson
					.fromJson<Map<String, MutableList<String>>>(
						obj,
						object : TypeToken<Map<String?, List<String?>?>?>() {}.type
					)
				this.data.clear()
				this.data.putAll(data)
			} else {
				log.warn(IbukiGourdLang.SetFromJsonFailed.tString(name, jsonElement))
			}
		} catch (e: Exception) {
			log.warn(IbukiGourdLang.SetFromJsonFailed.tString(name, jsonElement))
			log.error(e)
		}
	}

	override fun initialize() {

	}

	override val asJsonElement: JsonElement
		get() = JsonUtil.gson.toJsonTree(data)

	companion object {
		@Transient
		private val log = getLogger(Tooltip::class.java)
		fun getKey(stack: ItemStack): String {
			var value = ""
			val item = stack.item
			val name = item.getName(stack).string
			val cName = stack.name.string
			if (stack.item == Items.PLAYER_HEAD) {
				value = ":" + PlayerHeadUtil.getSkullOwner(stack)
			} else if (name != cName) {
				value = "#$cName"
			}
			return item.getTranslationKey(stack) + value
		}
	}
}